lib/logstash/outputs/elasticsearch.rb

```diff
@@ -367,36 +367,37 @@ def multi_receive(events)
     wait_for_successful_connection if @after_successful_connection_done
     events_mapped = safe_interpolation_map_events(events)
     retrying_submit(events_mapped.successful_events)
-    unless events_mapped.failed_events.empty?
-      @logger.error("Can't map some events, needs to be handled by DLQ #{events_mapped.failed_events}")
-      send_failed_resolutions_to_dlq(events_mapped.failed_events)
+    unless events_mapped.event_mapping_errors.empty?
+      handle_event_mapping_errors(events_mapped.event_mapping_errors)
     end
   end
 
-  # @param: Arrays of EventActionTuple
+  # @param: Arrays of EventActionTuple with error messages
   private
-  def send_failed_resolutions_to_dlq(failed_action_tuples)
-    failed_action_tuples.each do |action|
-      handle_dlq_status(action, "warn", "Could not resolve dynamic index")
+  def handle_event_mapping_errors(event_mapping_errors)
+    event_mapping_errors.each do |action, message|
+      detail_message = message + ", " + action.to_s
+      handle_dlq_status(action.event, :warn, detail_message)
+      @document_level_metrics.increment(:non_retryable_failures) unless @dlq_writer
     end
   end
 
-  MapEventsResult = Struct.new(:successful_events, :failed_events)
+  MapEventsResult = Struct.new(:successful_events, :event_mapping_errors)
 
   private
   def safe_interpolation_map_events(events)
     successful_events = [] # list of LogStash::Outputs::ElasticSearch::EventActionTuple
-    failed_events = [] # list of LogStash::Event
+    event_mapping_errors = [] # list of LogStash::Outputs::ElasticSearch::EventActionTuple with error messages
     events.each do |event|
-       begin
+      begin
         successful_events << @event_mapper.call(event)
-       rescue IndexInterpolationError, e
-         action = event.sprintf(@action || 'index')
-         event_action_tuple = EventActionTuple.new(action, [], event)
-         failed_events << event_action_tuple
+      rescue EventMappingError => ie
+        action = event.sprintf(@action || 'index')
+        event_action_tuple = EventActionTuple.new(action, [], event)
+        event_mapping_errors << [event_action_tuple, ie.message]
```

Comment:
@andsel , I have unified badly interpolated index name case with unsupported action. They both seem to me and Ry (as we quickly talked) common `EventMappingError` with different messages.
Also, as my understanding you used the `EventActionTuple` to send entire event body to DLQ. It seems we don't necessarily `sprintf` here and return `EventActionTuple` list, rather use the `Logstash::Event`.
I have tried `Logstash::Event` but I wasn't able to create detail message in L379. The reason `event.to_s` runs `Event.java -> toString()` and instead of inspecting entire data, it only considers `host` and `message` field. So for now, I have also aligned with `EventActionTuple` to properly send entire event data to DLQ (or drop by logging detail message). Let me know your opinion if we can still improve. Huge thanks for clear logic placement, BTW.

cc: @yaauie

Reply:
The motivation why I originally used `EventActionTuple` also for failed events, was that the original `#handle_dlq_status` required that as input argument, plus was kept also for type symmetry  in the two lists that composes `MapEventsResult(:successful_events, :failed_events)`.
I like the rephrase of `IndexInterpolationError` to inherit from `EventMappingError`.

Reply:
FYI: I have changed to Event as we don't need extra `sprintf`.